cmd/acr/purge.go

```diff
@@ -4,27 +4,222 @@
+
+Delete all dangling manifests
+  acr purge -r MyRegistry --repository MyRepository --dangling`
+)
+
+var (
+	registryName string
+	username     string
+	password     string
+	ago          string
+	dangling     bool
+	filter       string
+	repoName     string
 )
 
 func newPurgeCmd(out io.Writer) *cobra.Command {
 	cmd := &cobra.Command{
-		Use:   "purge",
-		Short: "", // TODO
-		Long:  purgeLongMessage,
+		Use:     "purge",
+		Short:   "Delete images from a registry.",
+		Long:    purgeLongMessage,
+		Example: exampleMessage,
 		RunE: func(cmd *cobra.Command, args []string) error {
-			fmt.Println("The default base uri is:", acrapi.DefaultBaseURI)
+			ctx := context.Background()
+			var e error
+			var wg sync.WaitGroup
+			loginURL := api.LoginURL(registryName)
+			auth := api.BasicAuth(username, password)
+			if !dangling {
+				e = PurgeTags(ctx, &wg, loginURL, auth, repoName)
+				if e != nil {
+					return e
+				}
+			}
+			e = PurgeDanglingManifests(ctx, &wg, loginURL, auth, repoName)
+			if e != nil {
+				return e
+			}
+
 			return nil
 		},
 	}
 
+	cmd.PersistentFlags().StringVarP(&registryName, "registry", "r", "", "Registry name")
+	cmd.MarkPersistentFlagRequired("registry")
+	cmd.PersistentFlags().StringVarP(&username, "username", "u", "", "Registry username")
+	cmd.MarkPersistentFlagRequired("username")
+	cmd.PersistentFlags().StringVarP(&password, "password", "p", "", "Registry password")
+	cmd.MarkPersistentFlagRequired("password")
+
+	cmd.Flags().StringVar(&ago, "ago", "1d", "The images that were created before this timeStamp will be deleted")
+	cmd.Flags().BoolVar(&dangling, "dangling", false, "Just remove dangling manifests")
+	cmd.Flags().StringVarP(&filter, "filter", "f", "", "Given as a regular expression, if a tag matches the pattern and is older than the time specified in ago it gets deleted.")
+	cmd.Flags().StringVar(&repoName, "repository", "", "The repository which will be purged.")
+	cmd.MarkFlagRequired("repository")
+
 	return cmd
 }
+
+// PurgeTags deletes all tags that are older than the ago value and that match the filter string (if present)
+func PurgeTags(ctx context.Context, wg *sync.WaitGroup, loginURL string, auth string, repoName string) error {
+	var days int
+	var durationString string
+	if strings.Contains(ago, "d") {
+		if _, e := fmt.Sscanf(ago, "%dd%s", &days, &durationString); e != nil {
+			fmt.Sscanf(ago, "%dd", &days)
+			durationString = ""
+		}
+	} else {
+		days = 0
+		if _, e := fmt.Sscanf(ago, "%s", &durationString); e != nil {
+			return e
+		}
+	}
+	timeToCompare := time.Now().UTC()
+	timeToCompare = timeToCompare.Add(time.Duration(-1*days) * 24 * time.Hour)
+	if len(durationString) > 0 {
+		agoDuration, e := time.ParseDuration(durationString)
+		if e != nil {
+			return e
+		}
+		timeToCompare = timeToCompare.Add(-1 * agoDuration)
+	}
+	var matches bool
+	var t time.Time
+	var errorChannel = make(chan error, 100)
```

Comment:
What if more than 100 tags for untagging?

Reply:
Because only 100 tags can be deleted in a loop iteration, then the channel gets emptied and then another 100 tags can be deleted in the next for iteration.